File: src/main/java/gargoyle/netsquares/prop/RWProperty.java
package gargoyle.netsquares.prop;

import gargoyle.netsquares.prop.base.BaseRWProperty;

import java.util.Objects;

public class RWProperty<T> extends BaseRWProperty<T> {
    private T value;

    public RWProperty(T value) {
        this.value = value;
    }

    public RWProperty() {
    }

    @Override
    public final T get() {
        return value;
    }

    @Override
    protected final void _set(T value) {
        this.value = value;
    }

    @Override
    public int hashCode() {
        return Objects.hash(super.hashCode(), value);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof RWProperty)) return false;
        if (!super.equals(o)) return false;
        RWProperty<?> that = (RWProperty<?>) o;
        return Objects.equals(value, that.value);
    }
}

File: src/main/kotlin/gargoyle/netsquares/prop/RWProperty.kt
package gargoyle.netsquares.prop

import gargoyle.netsquares.prop.base.BaseRWProperty

open class RWProperty<T : Any> : BaseRWProperty<T> {
    private lateinit var value: T

    constructor(value: T) {
        this.value = value
    }

    constructor()

    override fun _set(value: T) {
        this.value = value
    }

    override fun hashCode(): Int {
        return arrayOf(super.hashCode(), value).contentHashCode()
    }

    override fun get(): T {
        return value
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is RWProperty<*>) return false
        if (!super.equals(other)) return false
        return value == other.value
    }
}
